From e78c0a47e92da1e912a41958ef273afb970ad925 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Wed, 03 Mar 2021 15:33:58 +0200
Subject: [PATCH] [SW-20154] gaudi2_arc_gnu: impose checks on erom bins

As per recommendations from security team, ROM code
needs to make sure it performs DMA of known size
and within operating boundaries of flash. Hence the
eROM code is modified accordingly.

Change-Id: I94b5335b8775ccac4c520281a87775c84fd88c02
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/common/boot_desc.c b/common/boot_desc.c
index 359bd81..8095d80 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -14,7 +14,6 @@
 #include "dma.h"
 #include "spi_flash.h"
 
-#define MAX_IMGS		5
 #define BOOT_PTRS_SZ		sizeof(struct hl_fw_boot_ptrs)
 #define IMG_DESC_SZ		sizeof(struct hl_fw_img_desc)
 #define CRC32_PADD		sizeof(uint32_t)
@@ -22,8 +21,9 @@
 #define INV_VER1		((uint32_t)0)
 #define INV_VER2		((uint32_t) -1)
 #define INVALID_DESC_ID		2
+#define MAX_EROM_BINS		3
 static struct hl_fw_boot_ptrs *p_boot_ptrs;
-static struct hl_fw_img_desc *p_img_desc[MAX_IMGS];
+static struct hl_fw_img_desc *p_img_desc[MAX_EROM_BINS];
 
 static bool boot_desc_img_desc_is_valid(struct hl_fw_img_desc *img_desc)
 {
@@ -47,6 +47,10 @@
 
 	boot_ptrs_count = spi_flash_readl(HL_BOOT_PTS_COUNT_FLASH_OFFSET);
 
+	/* Expect just 'MAX_EROM_BINS' pointers, else give up */
+	if (boot_ptrs_count > MAX_EROM_BINS)
+		return 0;
+
 	img_ptrs_size = BOOT_PTRS_SZ + ((sizeof(uint32_t) +
 			IMG_DESC_SZ) * boot_ptrs_count) + CRC32_PADD;
 
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index 21e7725..c0cbb95 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -148,7 +148,18 @@
 	if (p_desc->img_size == 0)
 		goto img_desc_invalid;
 
-	if (p_desc->spi_offset + p_desc->img_size > FLASH_VALID_MASK)
+	/*
+	 * prevent rogue accesses by checking for-
+	 * 1. flash offset should be <= Max possible offset
+	 * 2. requested size is also less than Max allowed size
+	 */
+	if ((p_desc->spi_offset >=
+		(FLASH_VALID_OFFSET + FLASH_VALID_MAX_SIZE)) ||
+		(p_desc->img_size >= FLASH_VALID_MAX_SIZE))
+		goto img_desc_invalid;
+
+	if (p_desc->spi_offset + p_desc->img_size >=
+			(FLASH_VALID_OFFSET + FLASH_VALID_MAX_SIZE))
 		goto img_desc_invalid;
 
 	if (check_load_addr_valid(p_desc, addr))
diff --git a/gaudi2_arc_gnu/soc.h b/gaudi2_arc_gnu/soc.h
index 9c3af90..04a8713 100644
--- a/gaudi2_arc_gnu/soc.h
+++ b/gaudi2_arc_gnu/soc.h
@@ -19,6 +19,8 @@
 
 /* 2MB for Gaudi2 */
 #define FLASH_VALID_MASK		0x1FFFFF
+#define FLASH_VALID_OFFSET		0
+#define FLASH_VALID_MAX_SIZE		0x200000
 
 #define HL_SRAM_BASE			0x1000007FF0000000ULL
 #define HL_SP_SRAM_BASE			0x1000007FF7FE0000ULL
diff --git a/goya2_arc/soc.h b/goya2_arc/soc.h
index 55aac0e..8ef900d 100644
--- a/goya2_arc/soc.h
+++ b/goya2_arc/soc.h
@@ -23,6 +23,8 @@
 
 /* 16MB for Greco */
 #define FLASH_VALID_MASK			0xFFFFFF
+#define FLASH_VALID_OFFSET			0
+#define FLASH_VALID_MAX_SIZE			0x1000000
 
 #define HL_SRAM_BASE				0x7FF0000000ULL
 #define HL_SP_SRAM_BASE            		0x7FFBFE0000ULL
